### PR TITLE
fix document masking for chunked attention

### PR DESCRIPTION
**Pull request by winglian**
# What does this PR do?

The chunked attention was clobbering the document_ids meaning that the info to distinguish documents within the same chunk is gone. The proper way to do this is to simply generate a chunk mask and combine with the existing causal mask.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@ArthurZucker @SunMarc 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/integrations/flex_attention.py [hunk 1]
```diff
@@ -143,16 +143,26 @@ def causal_mask_mod(batch_idx, head_idx, q_idx, kv_idx):
         final_mask = causal_mask & padding_mask & document_mask
         return final_mask
 
+    def chunk_causal_mask_mod(batch_idx, head_idx, q_idx, kv_idx):
+        """
+        Combines the chunk mask with the causal mask for chunked attention.
+        """
+        chunk_mask = chunk_idxs[batch_idx, q_idx] == chunk_idxs[batch_idx, kv_idx]
+        causal_doc_mask = causal_mask_mod(batch_idx, head_idx, q_idx, kv_idx)
+        return chunk_mask & causal_doc_mask
```
Comment (ArthurZucker) [inline src/transformers/integrations/flex_attention.py:152]: we need the attention mask as well for padding and etc

**Review by ArthurZucker — COMMENTED** [review]
this does not affect l4 but uyes fixes the document id!

File: src/transformers/integrations/flex_attention.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:152]: The call to `causal_mask_mod` should handle this 🤔

Comment (ArthurZucker) [reply src/transformers/integrations/flex_attention.py:152]: Flex has the `mask_and` don't know if it's better or not.

**Review by ArthurZucker — APPROVED** [review]
Can you just make sure this works well for chunked attention decoding? Never know when compile can break? 😓

File: src/transformers/integrations/flex_attention.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:152]: They're basically doing the same in this case to my understanding. (Code is at https://github.com/pytorch/pytorch/blob/183bca41de89825107dce1d1ecc1502d9993a684/torch/nn/attention/flex_attention.py#L726-L730)

Comment (winglian) [reply src/transformers/integrations/flex_attention.py:152]: https://github.com/pytorch/torchtune/pull/2570/files#diff-6a071b9913a7be72d6547e874d9823c4eac4bc246c4e601d2179c4874ba06879R42-R48

touchtune relies pretty heavily on compile too

**Comment by winglian** [comment]
@ArthurZucker Anything I need to do to get this over the line?
